src/allmydata/test/test_istorageserver.py

```diff
@@ -0,0 +1,282 @@
+"""
+Tests for the ``IStorageServer`` interface.
+
+Note that for performance, in the future we might want the same node to be
+reused across tests, so each test should be careful to generate unique storage
+indexes.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from future.utils import PY2, bchr
+
+if PY2:
+    # fmt: off
+    from future.builtins import filter, map, zip, ascii, chr, hex, input, next, oct, open, pow, round, super, bytes, dict, list, object, range, str, max, min  # noqa: F401
+    # fmt: on
+
+from random import randrange
+
+from twisted.internet.defer import inlineCallbacks
+from twisted.trial.unittest import TestCase
+
+from foolscap.api import Referenceable
+
+from allmydata.interfaces import IStorageServer
+from .test_system import SystemTestMixin
+
+
+def _randbytes(length):
+    # type: (int) -> bytes
+    """Return random bytes string of given length."""
+    return b"".join([bchr(randrange(0, 256)) for _ in range(length)])
```

Comment:
`random.randrange` makes all of these tests non-reproduceable, does it not?  At least - if the behavior actually depends on the value for storage index (which it _mostly_ does not - except for collisions, probably?)

It would be lovely if we could use Hypothesis here ... but Hypothesis does not play nicely with AsyncTestCase so maybe that's not feasible.

Short of that, the tests could either:

* log all random values they use so it's at least possible to hand-craft a test run that reproduces a particular case
* hard-code a single storage index (and suppose that behavior really does not depend on its value)
* use a deterministic function to generate different values (like count up from 0) so they're different in each test but still easily predicted/known/reproduced

Also, in general, I'm really sad that Hypothesis and AsyncTestCase don't play well together because it seems like `allocate_buckets` is an excellent candidate for a Hypothesis stateful test which would automatically explore the state space more thoroughly and in much less code than it takes us to do it manually. :/  AFAICT this not an actionable comment but I couldn't not write it down. :(

Reply:
I switch to using `random.Random` with a known seed, so it should be reproducible so long as test order is the same.